### PR TITLE
[FLINK-13028][table] Move expression resolver to flink-table-api-java

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This finalizes the expression rework efforts by remove remaining dependencies to planner classes in the API module and refactors the package structure.

## Brief change log

See commit messages.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/ExpressionResolver.java [hunk 1]
```diff
@@ -97,15 +89,13 @@
 			ResolverRules.EXPAND_COLUMN_FUNCTIONS,
 			ResolverRules.OVER_WINDOWS,
 			ResolverRules.FIELD_RESOLVE,
-			ResolverRules.FLATTEN_CALL,
 			ResolverRules.QUALIFY_BUILT_IN_FUNCTIONS,
-			ResolverRules.RESOLVE_CALL_BY_ARGUMENTS);
+			ResolverRules.RESOLVE_CALL_BY_ARGUMENTS,
+			ResolverRules.FLATTEN_CALL);
```
Comment (dawidwys) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/ExpressionResolver.java]: I am not sure if it is correct to call the `FLATTEN_CALL` after `RESOLVE_CALL_BY_ARGUMENTS`. The `FLATTEN_CALL` may change the arguments list of other calls.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java [hunk 2]
```diff
@@ -38,75 +40,103 @@
...
+			}
+			throw new TableException("Unexpected unresolved expression: " + expression);
+		}
+
+		private List<ResolvedExpression> executeFlatten(List<ResolvedExpression> args) {
+			if (args.size() != 1) {
+				throw new ValidationException("Invalid number of arguments for flattening.");
+			}
+			final ResolvedExpression composite = args.get(0);
+			final TypeInformation<?> resultType = fromDataTypeToLegacyInfo(composite.getOutputDataType());
```
Comment (dawidwys) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java]: Shall we support also the new types?

Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java]: Structured types cannot be created yet. For row types it should work.

Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java]: Will it work if a row has e.g. `TIMESTAMP_WITH_TIMEZONE`? I agree though we don't necessarily have to do it immediately. Just wanted to point it out, as we port the type checks to the new type system in some places already.

Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java]: Yes you are right. I will leave a TODO there. We will need to touch this part of code again soon.

**Pull request closed without merging**